### PR TITLE
added maintenance exclusions to GKE

### PR DESCRIPTION
**Pull request by venkykuberan**
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
container : added maintenance_exclusions_window to `resource_google_container_cluster`

```
fixes https://github.com/hashicorp/terraform-provider-google/issues/7214

### CODE REVIEW

File: third_party/terraform/website/docs/r/container_cluster.html.markdown [hunk 1]
```diff
@@ -493,6 +493,37 @@ maintenance_policy {
 
 In beta, one or the other of `recurring_window` and `daily_maintenance_window` is required if a `maintenance_policy` block is supplied.
 
+* `maintenance_exclusions_window` - (Optional) Exceptions to maintenance window. Non-emergency maintenance should not occur in these windows. A cluster can have up to three maintenance exclusions at a time [Maintenance Window and Exclusions](https://cloud.google.com/kubernetes-engine/docs/concepts/maintenance-windows-and-exclusions)
+
+Specify `start_time` and `end_time` in [RFC3339](https://www.ietf.org/rfc/rfc3339.txt) "Zulu" date format.  The start time's date is
```
Comment (slevenick) [inline third_party/terraform/website/docs/r/container_cluster.html.markdown:498]: This format doesn't match others where we describe what a block supports, as seen below with:

The `ip_allocation_policy` block supports:

File: third_party/terraform/website/docs/r/container_cluster.html.markdown [hunk 2]
```diff
@@ -493,6 +493,37 @@ maintenance_policy {
...
+the initial date that the window starts, and the end time is used for calculating duration.  Specify `recurrence` in
+[RFC5545](https://tools.ietf.org/html/rfc5545#section-3.8.5.3) RRULE format, to specify when this recurs.
+Note that GKE may accept other formats, but will return values in UTC, causing a permanent diff.
+
+Examples:
```
Comment (slevenick) [inline third_party/terraform/website/docs/r/container_cluster.html.markdown:503]: Do we have examples inline for other fields in container cluster? This feels out of place

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 3]
```diff
@@ -2901,6 +2921,17 @@ func expandMaintenancePolicy(d *schema.ResourceData, meta interface{}) *containe
 	}
 	maintenancePolicy := l[0].(map[string]interface{})
 
+	if maintenanceExclusions, ok := maintenancePolicy["maintenance_exclusion_window"]; ok && len(maintenanceExclusions.(*schema.Set).List()) > 0 {
+		for k, me := range maintenanceExclusions.(*schema.Set).List() {
+			mew := me.(map[string]interface{})
+			blackout := fmt.Sprintf("blackout-%d", k+1)
+			exclusions[blackout] = containerBeta.TimeWindow{
```
Comment (slevenick) [inline third_party/terraform/resources/resource_container_cluster.go.erb:2928]: I feel like we should allow users to input the `name` of an exclusion window, as described here in the API guide: https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions#example-maintenance-exclusions

Can we add an `exclusion_name` or similar field to the exclusion object and use that as the key in the map here?

**Review by slevenick — CHANGES_REQUESTED** [review]

File: third_party/terraform/website/docs/r/container_cluster.html.markdown (hunk 2 above)
Comment (venkykuberan) [reply third_party/terraform/website/docs/r/container_cluster.html.markdown:503]: I see examples included for other attributes of 'maintenance_policy` here https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/container_cluster#maintenance_policy.

File: third_party/terraform/website/docs/r/container_cluster.html.markdown (hunk 1 above)
Comment (venkykuberan) [reply third_party/terraform/website/docs/r/container_cluster.html.markdown:498]: `maintenance_exclusion` is child attribute of `maintenance_policy` which is already in the format as
 
The `maintenance_policy` block supports:

Should i need to add similar format for `maintenance_exclusion` as well ?

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 4]
```diff
@@ -3567,6 +3601,18 @@ func flattenMaintenancePolicy(mp *containerBeta.MaintenancePolicy) []map[string]
 	if mp == nil || mp.Window == nil {
 		return nil
 	}
+
+	exclusionWindows := []map[string]interface{}{}
+	if mp.Window.MaintenanceExclusions != nil {
+		for wname, tw := range mp.Window.MaintenanceExclusions {
```
Comment (slevenick) [inline third_party/terraform/resources/resource_container_cluster.go.erb:3607]:
```suggestion
		for wName, tw := range mp.Window.MaintenanceExclusions {
```

Prefer camelCase for var names

Comment (slevenick) [inline third_party/terraform/resources/resource_container_cluster.go.erb:3607]: What does `tw` represent here? Maybe rename to `window` or similar?

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 5]
```diff
@@ -3567,6 +3601,18 @@ func flattenMaintenancePolicy(mp *containerBeta.MaintenancePolicy) []map[string]
 	if mp == nil || mp.Window == nil {
 		return nil
 	}
+
+	exclusionWindows := []map[string]interface{}{}
```
Comment (slevenick) [inline third_party/terraform/resources/resource_container_cluster.go.erb:3605]: `exclusions` or `maintenanceExclusions` rather than use `window` which is not used in the names of these objects?

File: third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 6]
```diff
@@ -3694,6 +3734,35 @@ resource "google_container_cluster" "with_recurring_maintenance_window" {
...
+		end_time = "%s"
+		recurrence = "FREQ=DAILY"
+	}
+	maintenance_exclusion {
+		exclusion_name = "batch job"
+		start_time = "%s"
+		end_time = "%s"
+	}
+	maintenance_exclusion {
+		exclusion_name = "hoilday data load"
```
Comment (slevenick) [inline third_party/terraform/tests/resource_container_cluster_test.go.erb:3757]:
```suggestion
		exclusion_name = "holiday data load"
```

Will this field accept spaces? Usually spaces don't work within map type objects

File: third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 7]
```diff
@@ -1434,6 +1434,46 @@ func TestAccContainerCluster_withRecurringMaintenanceWindow(t *testing.T) {
...
+			{
+				ResourceName:        resourceName,
+				ImportStateIdPrefix: "us-central1-a/",
+				ImportState:         true,
+				ImportStateVerify:   true,
+			},
+			{
+				Config: testAccContainerCluster_withMaintenanceExclusionWindow(cluster, "2020-01-01T00:00:00Z", "2020-01-02T00:00:00Z", "2020-05-01T00:00:00Z", "2020-05-02T00:00:00Z"),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckNoResourceAttr(resourceName,
```
Comment (slevenick) [inline third_party/terraform/tests/resource_container_cluster_test.go.erb:1463]: Why do we have this check? It doesn't seem to be related to this test?

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 8]
```diff
@@ -3576,6 +3622,7 @@ func flattenMaintenancePolicy(mp *containerBeta.MaintenancePolicy) []map[string]
 						"duration":   mp.Window.DailyMaintenanceWindow.Duration,
 					},
 				},
+				"maintenance_exclusion": exclusionWindows,
```
Comment (slevenick) [inline third_party/terraform/resources/resource_container_cluster.go.erb:3625]: Does maintenance exclusion apply to both daily and recurring windows? Can you one of both of the types of windows to the test?

File: third_party/terraform/website/docs/r/container_cluster.html.markdown (hunk 1 above)
Comment (slevenick) [reply third_party/terraform/website/docs/r/container_cluster.html.markdown:498]: Yeah I would. The current documentation of the fields is a bit confusing to me: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/container_cluster#daily_maintenance_window

It doesn't specify the sub types in the same way as other objects on that page

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 9]
```diff
@@ -2901,6 +2925,16 @@ func expandMaintenancePolicy(d *schema.ResourceData, meta interface{}) *containe
 	}
 	maintenancePolicy := l[0].(map[string]interface{})
 
+	if maintenanceExclusions, ok := maintenancePolicy["maintenance_exclusion"]; ok && len(maintenanceExclusions.(*schema.Set).List()) > 0 {
+		for _, me := range maintenanceExclusions.(*schema.Set).List() {
+			mew := me.(map[string]interface{})
```
Comment (slevenick) [inline third_party/terraform/resources/resource_container_cluster.go.erb:2930]: Prefer `exclusion` or similar
```suggestion
			exclusion := me.(map[string]interface{})
```

File: third_party/terraform/resources/resource_container_cluster.go.erb (hunk 4 above)
Comment (venkykuberan) [reply third_party/terraform/resources/resource_container_cluster.go.erb:3607]: corrected it

Comment (venkykuberan) [reply third_party/terraform/resources/resource_container_cluster.go.erb:3607]: tw - timewindow. I can change it to window

File: third_party/terraform/resources/resource_container_cluster.go.erb (hunk 5 above)
Comment (venkykuberan) [reply third_party/terraform/resources/resource_container_cluster.go.erb:3605]: changed to exclusions

File: third_party/terraform/resources/resource_container_cluster.go.erb (hunk 9 above)
Comment (venkykuberan) [reply third_party/terraform/resources/resource_container_cluster.go.erb:2930]: done

File: third_party/terraform/tests/resource_container_cluster_test.go.erb (hunk 7 above)
Comment (venkykuberan) [reply third_party/terraform/tests/resource_container_cluster_test.go.erb:1463]: agree, removed it

File: third_party/terraform/resources/resource_container_cluster.go.erb (hunk 8 above)
Comment (venkykuberan) [reply third_party/terraform/resources/resource_container_cluster.go.erb]: done

File: third_party/terraform/tests/resource_container_cluster_test.go.erb (hunk 6 above)
Comment (venkykuberan) [reply third_party/terraform/tests/resource_container_cluster_test.go.erb]: I see it in the Request/Response, i believe its accepting

File: third_party/terraform/website/docs/r/container_cluster.html.markdown [hunk 10]
```diff
@@ -449,11 +449,17 @@ The `authenticator_groups_config` block supports:
 * `security_group` - (Required) The name of the RBAC security group for use with Google security groups in Kubernetes RBAC. Group name must be in format `gke-security-groups@yourdomain.com`.
 
 The `maintenance_policy` block supports:
+* `daily_maintenance_window` - (Required in GA, Optional in Beta) structure documented below.
```
Comment (slevenick) [inline third_party/terraform/website/docs/r/container_cluster.html.markdown:452]: Is this still required in GA and optional in Beta? It seems like it functions the same way between beta and GA

File: third_party/terraform/website/docs/r/container_cluster.html.markdown [hunk 11]
```diff
@@ -491,7 +496,34 @@ maintenance_policy {
...
+the initial date that the window starts, and the end time is used for calculating duration.Specify `recurrence` in
+[RFC5545](https://tools.ietf.org/html/rfc5545#section-3.8.5.3) RRULE format, to specify when this recurs.
+Note that GKE may accept other formats, but will return values in UTC, causing a permanent diff. 
+
+Examples:
+
+```
+maintenance_policy {
+  recurring_window {
+		start_time = "2019-01-01T00:00:00Z"
```
Comment (slevenick) [inline third_party/terraform/website/docs/r/container_cluster.html.markdown:511]: Spacing here is different than below. Prefer consistent tabs and spaces

**Review by slevenick — APPROVED** [review]
LGTM once tests pass. Couple more minor notes

File: third_party/terraform/website/docs/r/container_cluster.html.markdown (hunk 11 above)
Comment (venkykuberan) [reply third_party/terraform/website/docs/r/container_cluster.html.markdown]: done
